cores/esp32/HardwareSerial.cpp

```diff
@@ -210,15 +228,17 @@ void HardwareSerial::_uartEventTask(void *args)
             if(xQueueReceive(uartEventQueue, (void * )&event, (portTickType)portMAX_DELAY)) {
                 switch(event.type) {
                     case UART_DATA:
-                        if(uart->_onReceiveCB && uart->available() > 0) uart->_onReceiveCB();
+                        if(uart->_onReceiveCB && uart->available() > 0  && event.timeout_flag) uart->_onReceiveCB();
                         break;
                     case UART_FIFO_OVF:
                         log_w("UART%d FIFO Overflow. Consider adding Hardware Flow Control to your Application.", uart->_uart_nr);
                         if(uart->_onReceiveErrorCB) uart->_onReceiveErrorCB(UART_FIFO_OVF_ERROR);
+                        xQueueReset(uartEventQueue);
```

Comment:
why lose all previous data?

Reply:
I did this because it was like this on the previous UART task (before latest SuGuiller commit). But maybe this could be ommited. The reason is that when you have a big overflow (overflowing multiple times the FIFO, the event queue gets filled with one UART_BUFFER_FULL and multiple UART_FIFO_OVF. I though it was redundant to be called multiple times for the same event.

Reply:
This is what I was talking about: https://github.com/espressif/arduino-esp32/blob/c014eaf3524dd45049a812a393c8fdfc92feb886/cores/esp32/esp32-hal-uart.c#L117

If you think this should be deleted, I have no opposition.

Reply:
I would rather discuss it, than just delete it :) @SuGlider want to pitch in an idea too?

Reply:
Well, I guess this is a bit relative. It could happen that you have a very fast overflow and you will be called multiple times with the UART_FIFO_OVF event. But thinking about this more thoroughly, maybe that is not a "normal" or "frequent" situation. I added it back because I saw it on the original esp32-hal-uart code and when I was testing the overflow with the serial monitor, I noticed that behaviour. But as I said, probably is not a frequent situation and you could eventually lose something important on the event queue? (maybe an UART_DATA event after an overflow event?). Probably this should be deleted. But I will let @SuGlider give this opinion about this. I would love to have this on 2.0.3 because I will be using that version as the new core for my project (2.0.2 has many bugs with UART).

Reply:
Last suggestion is about changing the name of the function ```void HardwareSerial::onReceiveTimeout(uint8_t symbols_timeout)``` to ```void HardwareSerial::setOnReceiveTimeout(uint8_t symbols_timeout)``` or some name like ```setRxTimeout()```.


Timeout mode shall not be forced in the code... it shall be an option. check 

``` cpp
//  from void HardwareSerial::onReceiveTimeout(uint8_t symbols_timeout) :::

    if(symbols_timeout == 0) {
        _onReceiveTimeout = 1; // Never disable timeout
    }
```

``` cpp
// from void HardwareSerial::begin(...) :::

    // Set UART RX timeout
    if (_uart != NULL && _onReceiveCB != NULL) {
        uart_set_rx_timeout(_uart_nr, _onReceiveTimeout);
    }
```

``` cpp
// from void HardwareSerial::_uartEventTask(void *args)) ::: "&& event.timeout_flag"

                    case UART_DATA:
                        if(uart->_onReceiveCB && uart->available() > 0  && event.timeout_flag) uart->_onReceiveCB();
                        break;
```